### DISCUSSION TITLE
neovim cmp intergration for props and emits

### DISCUSSION THREAD

**Discussion by RayGuo-ergou** [vuejs/language-tools#4495] [General]
Hi neovim users, I want to introduce a `nvim-cmp` integration I am using for props and emits inspired by [this post](https://twitter.com/youyuxi/status/1799701740000035136).

Basically changes:

![image](https://github.com/vuejs/language-tools/assets/33137074/d5486a58-85ca-431f-b53c-298ed9622e3f)
![image](https://github.com/vuejs/language-tools/assets/33137074/71b65836-e3cb-4970-9ef7-3640f24394d5)

to:

![image](https://github.com/vuejs/language-tools/assets/33137074/13256ddb-1b85-4787-8380-99a5f2f95542)
![image](https://github.com/vuejs/language-tools/assets/33137074/92ade4e7-9839-4ded-ba00-ec7e81905786)

<hr />
Pass this into your `nvim-cmp` setup opts:

```lua
-- To see more information :h cmp-config.sources
sources = cmp.config.sources({
  {
    name = 'nvim_lsp',
    ---@param entry cmp.Entry
    ---@param ctx cmp.Context
    entry_filter = function(entry, ctx)
      -- Check if the buffer type is 'vue'
      if ctx.filetype ~= 'vue' then
        return true
      end

      local cursor_before_line = ctx.cursor_before_line
      -- For events
      if cursor_before_line:sub(-1) == '@' then
        return entry.completion_item.label:match('^@')
      -- For props also exclude events with `:on-` prefix
      elseif cursor_before_line:sub(-1) == ':' then
        return entry.completion_item.label:match('^:') and not entry.completion_item.label:match('^:on%-')
      else
        return true
      end
    end 
  },
})

```
The regex performance is quite good since it only checks the beginning characters. However, this approach can still lead to issues such as a string union.

For example, with this TypeScript type:

```ts
type EmailAddresses = 'a@b.c' | 'd@e.f'

```
You will lose the items in nvim-cmp if you only type `@`. To address this, you can add Treesitter to check if the cursor is in a starting tag. ( I am not a fan of complex regex )

To use a local buffer variable to cache the result of the Treesitter is recommended to avoid performance issues as the function will call for every single entry. Here is an example of how to do it:

```lua
-- check if in start tag
local function is_in_start_tag()
  local node = vim.treesitter.get_node()
  if not node then
    return false
  end
  local node_to_check = { 'start_tag', 'self_closing_tag', 'directive_attribute' }
  return vim.tbl_contains(node_to_check, node:type())
end


entry_filter = function(entry, ctx)
  -- Use a buffer-local variable to cache the result of the Treesitter check
  local bufnr = ctx.bufnr
  local cached_is_in_start_tag = vim.b[bufnr]._vue_ts_cached_is_in_start_tag
  if cached_is_in_start_tag == nil then
    vim.b[bufnr]._vue_ts_cached_is_in_start_tag = is_in_start_tag()
  end
  -- If not in start tag, return true
  if vim.b[bufnr]._vue_ts_cached_is_in_start_tag == false then
    return true
  end
  -- rest of the code
end

```

In your cmp configuration, register an event to clear the cache when the menu is closed:

```lua
cmp.event:on('menu_closed', function()
  local bufnr = vim.api.nvim_get_current_buf()
  vim.b[bufnr]._vue_ts_cached_is_in_start_tag = nil
end)
```

**Comment by cjrh** [comment]
There's an `end` missing for the function:

```lua
-- To see more information :h cmp-config.sources
sources = cmp.config.sources({
  {
    name = 'nvim_lsp',
    ---@param entry cmp.Entry
    ---@param ctx cmp.Context
    entry_filter = function(entry, ctx)
      -- Check if the buffer type is 'vue'
      if ctx.filetype ~= 'vue' then
        return true
      end

      local cursor_before_line = ctx.cursor_before_line
      -- For events
      if cursor_before_line:sub(-1) == '@' then
        return entry.completion_item.label:match('^@')
      -- For props also exclude events with `:on-` prefix
      elseif cursor_before_line:sub(-1) == ':' then
        return entry.completion_item.label:match('^:') and not entry.completion_item.label:match('^:on%-')
      else
        return true
      end
    end   -- <--- HERE
  },
})
```

**Reply by RayGuo-ergou** [reply]
Thanks! I have upated

**Comment by RayGuo-ergou** [comment]
I found that `start_tag` is not the only type of node I want this filter so the new function for treesitter is:

```lua
local function is_in_start_tag()
  local ts_utils = require('nvim-treesitter.ts_utils')
  local node = ts_utils.get_node_at_cursor()
  if not node then
    return false
  end
  local node_to_check = { 'start_tag', 'self_closing_tag', 'directive_attribute' }
  return vim.tbl_contains(node_to_check, node:type())
end
```

**Comment by exzolink** [comment]
It seems that after the latest updates, it no longer works.

@RayGuo-ergou can you confirm?

**Reply by RayGuo-ergou** [reply]
Still working for me. the only noticeable change is `:on-vnode-*` changed to `@vnode*`

**Reply by exzolink** [reply]
Updated to 2.1.4, now it works.

**Comment by RayGuo-ergou** [comment]
Made a small update to use neovim native api to get TS node, here's the diff. From what i have tested so far, it works identical. (Maybe require neovim version ^0.11 not sure)

```diff
-  local ts_utils = require('nvim-treesitter.ts_utils')
-  local node = ts_utils.get_node_at_cursor()
+  local node = vim.treesitter.get_node()
```

see `:h vim.treesitter.get_node()`
